src/Workspaces/CoreTest/UtilityTest/BKTreeTests.cs

```diff
@@ -0,0 +1,151 @@
+                }
+            }
+
+            foreach (var value in testValues)
+            {
+                // If we add a random character at any location in a string, we should still 
+                // be able to find it.
+                for (var i = 0; i <= value.Length; i++)
+                {
+                    var items = tree.Find(Insert(value, i, 'Z'), threshold: null);
+                    Assert.Contains(value.ToLower(), items);
+
+                    // We better not be finding all items.
+                    Assert.NotEqual(testValues.Length, items.Count);
+                }
+            }
+
+            foreach (var value in testValues)
+            {
+                // If we transpose any characters in a string, we should still 
+                // be able to find it.
+                for (var i = 0; i < value.Length - 1; i++)
+                {
+                    var items = tree.Find(Transpose(value, i), threshold: null);
+                    Assert.Contains(value.ToLower(), items);
+                }
+            }
+        }
+
+        private string Transpose(string value, int i)
+        {
+            return value.Substring(0, i) + value[i + 1] + value[i] + value.Substring(i + 2);
+        }
+
+        private string Insert(string value, int i, char v)
+        {
+            return value.Substring(0, i) + v + value.Substring(i);
+        }
+
+        private string Delete(string value, int i)
+        {
+            return value.Substring(0, i) + value.Substring(i + 1);
+        }
+
+        [Fact]
+        public void Test2()
+        {
+            string[] testValues = { "Leeds", "York", "Bristol", "Leicester", "Hull", "Durham" };
+            var tree = BKTree.Create(testValues);
+
+            var results = tree.Find("hill", threshold: null);
+            Assert.True(results.SetEquals(Expected("hull")));
+
+            results = tree.Find("liecester", threshold: null);
+            Assert.True(results.SetEquals(Expected("leicester")));
+
+            results = tree.Find("leicestre", threshold: null);
+            Assert.True(results.SetEquals(Expected("leicester")));
+
+            results = tree.Find("lecester", threshold: null);
+            Assert.True(results.SetEquals(Expected("leicester")));
+        }
+
+        [Fact]
+        public void TestSpillover()
+        {
+            string[] testValues = {
+                /*root:*/ "Four",
+                /*d=1*/ "Fou", "For", "Fur", "Our", "FourA", "FouAr", "FoAur", "FAour", "AFour", "Tour",
+                /*d=2*/ "Fo", "Fu", "Fr", "or", "ur", "ou", "FourAb", "FouAbr", "FoAbur", "FAbour", "AbFour", "oFour", "Fuor", "Foru", "ours",
+                /*d=3*/ "F", "o", "u", "r", "Fob", "Fox", "bur", "urn", "hur", "foraa", "found"
+            };
+            TestTreeInvariants(testValues);
+        }
+
+        [Fact]
+        public void Top1000()
+        {
+            TestTreeInvariants(EditDistanceTests.Top1000);
```

Comment:
Are there many things in the top 1000 that are within edit distance 1 of each other? Maybe do larger edit distance testing for this case.